File: http/java/okhttp/lib/src/main/java/com/microsoft/kiota/http/middleware/RetryHandler.java
package com.microsoft.kiota.http.middleware;

import java.io.IOException;
import java.util.Objects;

import javax.annotation.Nullable;
import javax.annotation.Nonnull;

import com.microsoft.kiota.http.middleware.IShouldRetry;
import com.microsoft.kiota.http.middleware.MiddlewareType;
import com.microsoft.kiota.http.middleware.options.RetryOptions;
import com.microsoft.kiota.http.middleware.options.TelemetryOptions;
import com.microsoft.kiota.http.logger.DefaultLogger;
import com.microsoft.kiota.http.logger.ILogger;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;

/**
 * The middleware responsible for retrying requests when they fail because of transient issues
 */
public class RetryHandler implements Interceptor{

    /**
     * Type of the current middleware
     */
    public final MiddlewareType MIDDLEWARE_TYPE = MiddlewareType.RETRY;

    private RetryOptions mRetryOption;

    /**
     * Header name to track the retry attempt number
     */
    private static final String RETRY_ATTEMPT_HEADER = "Retry-Attempt";
    /**
     * Header name for the retry after information
     */
    private static final String RETRY_AFTER = "Retry-After";

    /**
     * Too many requests status code
     */
    public static final int MSClientErrorCodeTooManyRequests = 429;
    /**
     * Service unavailable status code
     */
    public static final int MSClientErrorCodeServiceUnavailable  = 503;
    /**
     * Gateway timeout status code
     */
    public static final int MSClientErrorCodeGatewayTimeout = 504;

    /**
     * One second as milliseconds
     */
    private static final long DELAY_MILLISECONDS = 1000;

    private final ILogger logger;

    /**
     * @param retryOption Create Retry handler using retry option
     */
    public RetryHandler(@Nullable final RetryOptions retryOption) {
        this(new DefaultLogger(), retryOption);
    }
    /**
     * @param retryOption Create Retry handler using retry option
     * @param logger logger to use for telemetry
     */
    public RetryHandler(@Nonnull final ILogger logger, @Nullable final RetryOptions retryOption) {
        this.logger = Objects.requireNonNull(logger, "logger parameter cannot be null");
        this.mRetryOption = retryOption;
        if(this.mRetryOption == null) {
            this.mRetryOption = new RetryOptions();
        }
    }
    /**
     * Initialize retry handler with default retry option
     */
    public RetryHandler() {
        this(null);
    }

    boolean retryRequest(Response response, int executionCount, Request request, RetryOptions retryOptions) {

        // Should retry option
        // Use should retry common for all requests
        IShouldRetry shouldRetryCallback = null;
        if(retryOptions != null) {
            shouldRetryCallback = retryOptions.shouldRetry();
        }

        boolean shouldRetry = false;
        // Status codes 429 503 504
        int statusCode = response.code();
        // Only requests with payloads that are buffered/rewindable are supported.
        // Payloads with forward only streams will be have the responses returned
        // without any retry attempt.
        shouldRetry =
                retryOptions != null
                        && executionCount <= retryOptions.maxRetries()
                        && checkStatus(statusCode) && isBuffered(request)
                        && shouldRetryCallback != null
                        && shouldRetryCallback.shouldRetry(retryOptions.delay(), executionCount, request, response);

        if(shouldRetry) {
            long retryInterval = getRetryAfter(response, retryOptions.delay(), executionCount);
            try {
                Thread.sleep(retryInterval);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                logger.logError("error retrying the request", e);
            }
        }
        return shouldRetry;
    }

    /**
     * Get retry after in milliseconds
     * @param response Response
     * @param delay Delay in seconds
     * @param executionCount Execution count of retries
     * @return Retry interval in milliseconds
     */
    long getRetryAfter(Response response, long delay, int executionCount) {
        String retryAfterHeader = response.header(RETRY_AFTER);
        double retryDelay = RetryOptions.DEFAULT_DELAY * DELAY_MILLISECONDS;
        if(retryAfterHeader != null) {
            retryDelay = Long.parseLong(retryAfterHeader) * DELAY_MILLISECONDS;
        } else {
            retryDelay = (double)((Math.pow(2.0, (double)executionCount)-1)*0.5);
            retryDelay = (executionCount < 2 ? delay : retryDelay + delay) + (double)Math.random();
            retryDelay *= DELAY_MILLISECONDS;
        }
        return (long)Math.min(retryDelay, RetryOptions.MAX_DELAY * DELAY_MILLISECONDS);
    }

    boolean checkStatus(int statusCode) {
        return (statusCode == MSClientErrorCodeTooManyRequests || statusCode == MSClientErrorCodeServiceUnavailable
                || statusCode == MSClientErrorCodeGatewayTimeout);
    }

    boolean isBuffered(final Request request) {
        final String methodName = request.method();

        final boolean isHTTPMethodPutPatchOrPost = methodName.equalsIgnoreCase("POST") ||
                methodName.equalsIgnoreCase("PUT") ||
                methodName.equalsIgnoreCase("PATCH");

        final RequestBody requestBody = request.body();
        if(isHTTPMethodPutPatchOrPost && requestBody != null) {
            try {
                return requestBody.contentLength() != -1L;
            } catch (IOException ex) {
                // expected
                return false;
            }
        }
        return true;
    }

    @Override
    @Nonnull
    public Response intercept(@Nonnull final Chain chain) throws IOException {
        Request request = chain.request();

        TelemetryOptions telemetryOptions = request.tag(TelemetryOptions.class);
        if(telemetryOptions == null) {
            telemetryOptions = new TelemetryOptions();
            request = request.newBuilder().tag(TelemetryOptions.class, telemetryOptions).build();
        }
        telemetryOptions.setFeatureUsage(TelemetryOptions.RETRY_HANDLER_ENABLED_FLAG);

        Response response = chain.proceed(request);

        // Use should retry pass along with this request
        RetryOptions retryOption = request.tag(RetryOptions.class);
        retryOption = retryOption != null ? retryOption : mRetryOption;

        int executionCount = 1;
        while(retryRequest(response, executionCount, request, retryOption)) {
            request = request.newBuilder().addHeader(RETRY_ATTEMPT_HEADER, String.valueOf(executionCount)).build();
            executionCount++;
            if(response != null) {
                final ResponseBody body = response.body();
                if(body != null)
                    body.close();
                response.close();
            }
            response = chain.proceed(request);
        }
        return response;
    }

}

File: http/typescript/fetch/src/middlewares/retryHandler.ts
/**
 * -------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.
 * See License in the project root for license information.
 * -------------------------------------------------------------------------------------------
 */

/**
 * @module RetryHandler
 */

import { HttpMethod, RequestOption } from "@microsoft/kiota-abstractions";

import { FetchRequestInit, FetchResponse } from "../utils/fetchDefinitions";
import { getRequestHeader, setRequestHeader } from "../utils/headersUtil";
import { Middleware } from "./middleware";
import { RetryHandlerOptionKey, RetryHandlerOptions } from "./options/retryHandlerOptions";

/**
 * @class
 * @implements Middleware
 * Class for RetryHandler
 */
export class RetryHandler implements Middleware {
	/**
	 * @private
	 * @static
	 * A list of status codes that needs to be retried
	 */
	private static RETRY_STATUS_CODES: Set<number> = new Set([
		429, // Too many requests
		503, // Service unavailable
		504, // Gateway timeout
	]);

	/**
	 * @private
	 * @static
	 * A member holding the name of retry attempt header
	 */
	private static RETRY_ATTEMPT_HEADER = "Retry-Attempt";

	/**
	 * @private
	 * @static
	 * A member holding the name of retry after header
	 */
	private static RETRY_AFTER_HEADER = "Retry-After";

	/**
	 * @private
	 * The next middleware in the middleware chain
	 */
	next: Middleware | undefined;

	/**
	 * @public
	 * @constructor
	 * To create an instance of RetryHandler
	 * @param {RetryHandlerOptions} [options = new RetryHandlerOptions()] - The retry handler options value
	 * @returns An instance of RetryHandler
	 */
	public constructor(private options: RetryHandlerOptions = new RetryHandlerOptions()) {}

	/**
	 *
	 * @private
	 * To check whether the response has the retry status code
	 * @param {Response} response - The response object
	 * @returns Whether the response has retry status code or not
	 */
	private isRetry(response: FetchResponse): boolean {
		return RetryHandler.RETRY_STATUS_CODES.has(response.status);
	}

	/**
	 * @private
	 * To check whether the payload is buffered or not
	 * @param {RequestInit} options - The options of a request
	 * @returns Whether the payload is buffered or not
	 */
	private isBuffered(options: FetchRequestInit): boolean {
		const method = options.method;
		const isPutPatchOrPost: boolean = method === HttpMethod.PUT || method === HttpMethod.PATCH || method === HttpMethod.POST;
		if (isPutPatchOrPost) {
			const isStream = getRequestHeader(options, "content-type")?.toLowerCase() === "application/octet-stream";
			if (isStream) {
				return false;
			}
		}
		return true;
	}

	/**
	 * @private
	 * To get the delay for a retry
	 * @param {Response} response - The response object
	 * @param {number} retryAttempts - The current attempt count
	 * @param {number} delay - The delay value in seconds
	 * @returns A delay for a retry
	 */
	private getDelay(response: FetchResponse, retryAttempts: number, delay: number): number {
		const getRandomness = () => Number(Math.random().toFixed(3));
		const retryAfter = response.headers !== undefined ? response.headers.get(RetryHandler.RETRY_AFTER_HEADER) : null;
		let newDelay: number;
		if (retryAfter !== null) {
			// Retry-After: <http-date>
			if (Number.isNaN(Number(retryAfter))) {
				newDelay = Math.round((new Date(retryAfter).getTime() - Date.now()) / 1000);
			} else {
				// Retry-After: <delay-seconds>
				newDelay = Number(retryAfter);
			}
		} else {
			// Adding randomness to avoid retrying at a same
			newDelay = retryAttempts >= 2 ? this.getExponentialBackOffTime(retryAttempts) + delay + getRandomness() : delay + getRandomness();
		}
		return Math.min(newDelay, this.options.getMaxDelay() + getRandomness());
	}

	/**
	 * @private
	 * To get an exponential back off value
	 * @param {number} attempts - The current attempt count
	 * @returns An exponential back off value
	 */
	private getExponentialBackOffTime(attempts: number): number {
		return Math.round((1 / 2) * (2 ** attempts - 1));
	}

	/**
	 * @private
	 * @async
	 * To add delay for the execution
	 * @param {number} delaySeconds - The delay value in seconds
	 * @returns Nothing
	 */
	private async sleep(delaySeconds: number): Promise<void> {
		const delayMilliseconds = delaySeconds * 1000;
		return new Promise((resolve) => setTimeout(resolve, delayMilliseconds)); // browser or node
	}

	/**
	 * @private
	 * @async
	 * To execute the middleware with retries
	 * @param {Context} context - The context object
	 * @param {number} retryAttempts - The current attempt count
	 * @param {RetryHandlerOptions} options - The retry middleware options instance
	 * @returns A Promise that resolves to nothing
	 */
	private async executeWithRetry(url: string, fetchRequestInit: FetchRequestInit, retryAttempts: number, requestOptions?: Record<string, RequestOption>): Promise<FetchResponse> {
		const response = await this.next?.execute(url, fetchRequestInit as RequestInit, requestOptions);
		if (!response) {
			throw new Error("Response is undefined");
		}
		if (retryAttempts < this.options.maxRetries && this.isRetry(response!) && this.isBuffered(fetchRequestInit) && this.options.shouldRetry(this.options.delay, retryAttempts, url, fetchRequestInit! as RequestInit, response)) {
			++retryAttempts;
			setRequestHeader(fetchRequestInit, RetryHandler.RETRY_ATTEMPT_HEADER, retryAttempts.toString());
			if (response) {
				const delay = this.getDelay(response!, retryAttempts, this.options.delay);
				await this.sleep(delay);
			}
			return await this.executeWithRetry(url, fetchRequestInit, retryAttempts, requestOptions);
		} else {
			return response;
		}
	}

	/**
	 * @public
	 * @async
	 * To execute the current middleware
	 * @param {Context} context - The context object of the request
	 * @returns A Promise that resolves to nothing
	 */
	public execute(url: string, requestInit: RequestInit, requestOptions?: Record<string, RequestOption>): Promise<Response> {
		const retryAttempts = 0;

		if (requestOptions && requestOptions[RetryHandlerOptionKey]) {
			this.options = requestOptions[RetryHandlerOptionKey] as RetryHandlerOptions;
		}
		return this.executeWithRetry(url, requestInit as FetchRequestInit, retryAttempts, requestOptions);
	}
}
